### PR TITLE
feat(multi-trie): Added support for persistent storage tries

### PR DESCRIPTION
**Pull request by antiyro**
added support for persistent storage tries

fixed block 0 hash computation

### CODE REVIEW

File: crates/client/sync/src/commitments/lib.rs [hunk 1]
```diff
@@ -198,29 +206,46 @@ fn contract_trie_root<B: BlockT>(
...
+fn contract_state_leaf_hash<B: BlockT>(
     csd: &CommitmentStateDiff,
     overrides: &OverrideHandle<Block<Header<u32, BlakeTwo256>, OpaqueExtrinsic>>,
     contract_address: &ContractAddress,
-    storage_updates: &IndexMap<StorageKey, StarkFelt>,
     maybe_block_hash: Option<H256>,
+    bonsai_contract_storage: &Arc<Mutex<BonsaiStorage<BasicId, BonsaiDb<B>, Pedersen>>>,
 ) -> Result<Felt252Wrapper, BonsaiStorageError<BonsaiDbError>> {
-    let storage_root = update_storage_trie(overrides, contract_address, storage_updates, maybe_block_hash)?;
+    let identifier = contract_address.0.0.0.as_bytes_ref();
```
Comment (Trantorian1) [inline crates/client/sync/src/commitments/lib.rs]: Maybe consider refactoring this into an `identifier` function, similar to `key`?

File: crates/client/sync/src/commitments/lib.rs [hunk 2]
```diff
@@ -266,15 +291,19 @@ fn class_trie_root<B: BlockT>(
     mut bonsai_class: MutexGuard<BonsaiStorage<BasicId, BonsaiDb<B>, Poseidon>>,
     block_number: u64,
 ) -> Result<Felt252Wrapper, BonsaiStorageError<BonsaiDbError>> {
+    let identifier = "0xclass".as_bytes();
```
Comment (Trantorian1) [inline crates/client/sync/src/commitments/lib.rs]: `"0xclass"` should be a `pub const`

File: crates/client/sync/src/commitments/lib.rs [hunk 3]
```diff
@@ -198,29 +206,46 @@ fn contract_trie_root<B: BlockT>(
     csd: &CommitmentStateDiff,
     overrides: Arc<OverrideHandle<Block<Header<u32, BlakeTwo256>, OpaqueExtrinsic>>>,
     mut bonsai_contract: MutexGuard<BonsaiStorage<BasicId, BonsaiDb<B>, Pedersen>>,
+    bonsai_contract_storage: Arc<Mutex<BonsaiStorage<BasicId, BonsaiDb<B>, Pedersen>>>,
     block_number: u64,
     maybe_block_hash: Option<H256>,
 ) -> Result<Felt252Wrapper, BonsaiStorageError<BonsaiDbError>> {
+    let identifier = "0xcontract".as_bytes();
```
Comment (Trantorian1) [inline crates/client/sync/src/commitments/lib.rs]: `"0xcontract"` should be a `pub const`

File: crates/client/sync/src/commitments/transactions.rs [hunk 4]
```diff
@@ -69,10 +69,12 @@ pub async fn memory_transaction_commitment(
     chain_id: Felt252Wrapper,
     block_number: u64,
 ) -> Result<Felt252Wrapper, String> {
+    // TODO @cchudant refacto/optimise this function
     let config = BonsaiStorageConfig::default();
     let bonsai_db = HashMapDb::<BasicId>::default();
     let mut bonsai_storage =
         BonsaiStorage::<_, _, Pedersen>::new(bonsai_db, config).expect("Failed to create bonsai storage");
+    let identifier = "0xtransaction".as_bytes();
```
Comment (Trantorian1) [inline crates/client/sync/src/commitments/transactions.rs]: `"0xtransaction"` should be a `pub const`

File: crates/client/sync/src/l2.rs [hunk 5]
```diff
@@ -480,10 +480,12 @@ pub fn verify_l2<B: BlockT>(
         csd,
         Arc::clone(overrides),
         Arc::clone(bonsai_contract),
+        Arc::clone(bonsai_contract_storage),
         Arc::clone(bonsai_class),
         block_number,
         substrate_block_hash,
     );
+    println!("state_root: {state_root:?}");
```
Comment (Trantorian1) [inline crates/client/sync/src/l2.rs]: Please remove this or replace it by a `log::debug`

File: crates/primitives/block/src/header.rs [hunk 6]
```diff
@@ -159,11 +159,38 @@ impl Header {
...
 }
+
+#[cfg(test)]
+mod tests {
+    use mp_hashers::pedersen::PedersenHasher;
+
+    use super::*;
+
+    #[test]
+    fn test_header_hash() {
```
Comment (Trantorian1) [inline crates/primitives/block/src/header.rs]: Please consider documenting this test and it's purpose

File: crates/client/sync/src/commitments/contracts.rs [hunk 7]
```diff
@@ -39,43 +36,19 @@ pub struct ContractLeafParams {
...
-                if value != Felt252Wrapper::ZERO {
-                    bonsai_storage.insert(&key, &value.into()).expect("Failed to insert storage update into trie");
-                }
-            });
-        }
-    }
+    bonsai_contract_storage: &Arc<Mutex<BonsaiStorage<BasicId, BonsaiDb<B>, Pedersen>>>,
+) {
+    let mut bonsai_storage = bonsai_contract_storage.lock().unwrap();
+    let identifier = contract_address.0.0.0.as_bytes_ref();
```
Comment (Trantorian1) [inline crates/client/sync/src/commitments/contracts.rs]: Please refactor this to a public `identifier` function, similar to `key`

**Review by Trantorian1 — CHANGES_REQUESTED** [review]

File: crates/primitives/block/src/header.rs (hunk 6 above)
Comment (antiyro) [reply crates/primitives/block/src/header.rs]: removed as it is useless for now

**Review by jbcaron — APPROVED** [review]

**Review by Trantorian1 — APPROVED** [review]

**Pull request merged**